libkbfs/tlf_journal.go

```diff
@@ -402,33 +416,134 @@ func (j *tlfJournal) flush(ctx context.Context) (err error) {
+		journalOrdinal, *blockJournalEntry, []byte,
+		BlockCryptKeyServerHalf, error) {
+		j.journalLock.RLock()
+		defer j.journalLock.RUnlock()
+		return j.blockJournal.getNextEntryToFlush(ctx)
+	}()
+	if err != nil {
+		return false, err
+	}
+	if e == nil {
+		return false, nil
+	}
+
+	err = flushBlockJournalEntry(
+		ctx, j.log, j.delegateBlockServer, j.tlfID, *e,
+		data, serverHalf)
+	if err != nil {
+		return false, err
+	}
+
+	// We're the only thing removing from the block journal, so we
+	// can assume that the earliest op is the one we just got.
+	j.journalLock.Lock()
+	defer j.journalLock.Unlock()
+	err = j.blockJournal.removeFlushedEntry(ctx, o, *e)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
 }
 
 func (j *tlfJournal) flushOneMDOp(ctx context.Context) (bool, error) {
-	_, currentUID, err := j.config.KBPKI().GetCurrentUserInfo(ctx)
+	uid, key, err :=
+		getCurrentUIDAndVerifyingKey(ctx, j.config.currentInfoGetter())
 	if err != nil {
 		return false, err
 	}
 
-	currentVerifyingKey, err := j.config.KBPKI().GetCurrentVerifyingKey(ctx)
+	j.log.CDebugf(ctx, "Flushing one MD to server")
+	defer func() {
+		if err != nil {
+			j.deferLog.CDebugf(ctx, "Flush failed with %v", err)
+		}
+	}()
+
+	j.flushLock.Lock()
+	defer j.flushLock.Unlock()
+
+	signer := j.config.Crypto()
+	mdServer := j.config.MDServer()
+
+	mdID, rmds, err := func() (MdID, *RootMetadataSigned, error) {
+		j.journalLock.Lock()
+		defer j.journalLock.Unlock()
+		return j.mdJournal.getNextEntryToFlush(ctx, uid, key, signer)
+	}()
 	if err != nil {
 		return false, err
 	}
+	if mdID == (MdID{}) {
+		return false, nil
+	}
 
-	j.lock.Lock()
-	defer j.lock.Unlock()
-	return j.mdJournal.flushOne(
-		ctx, currentUID, currentVerifyingKey, j.config.Crypto(),
-		j.config.MDServer())
+	j.log.CDebugf(ctx, "Flushing MD for TLF=%s with id=%s, rev=%s, bid=%s",
+		rmds.MD.TlfID(), mdID, rmds.MD.RevisionNumber(), rmds.MD.BID())
+	pushErr := mdServer.Put(ctx, rmds)
+	if isRevisionConflict(pushErr) {
+		headMdID, err := getMdID(
+			ctx, mdServer, j.mdJournal.crypto, rmds.MD.TlfID(), rmds.MD.BID(),
+			rmds.MD.MergedStatus(), rmds.MD.RevisionNumber())
+		if err != nil {
+			j.log.CWarningf(ctx,
+				"getMdID failed for TLF %s, BID %s, and revision %d: %v",
+				rmds.MD.TlfID(), rmds.MD.BID(), rmds.MD.RevisionNumber(), err)
+		} else if headMdID == mdID {
+			if headMdID == (MdID{}) {
+				panic("nil earliestID and revision conflict error returned by pushEarliestToServer")
+			}
+			// We must have already flushed this MD, so continue.
+			pushErr = nil
+		} else if rmds.MD.MergedStatus() == Merged {
+			j.log.CDebugf(ctx, "Conflict detected %v", pushErr)
+
+			mdID, rmds, err = func() (MdID, *RootMetadataSigned, error) {
+				j.journalLock.Lock()
+				defer j.journalLock.Unlock()
+				err := j.mdJournal.convertToBranch(
+					ctx, uid, key, signer)
+				if err != nil {
+					return MdID{}, nil, err
+				}
+
+				return j.mdJournal.getNextEntryToFlush(
```

Comment:
Yes, we're unstaging all the MDs in the journal (`convertToBranch`) and trying again.